Source/ExecutionEngine/ImplementationTask.cs

```diff
@@ -62,27 +61,22 @@ public class ImplementationTask : IImplementationTask {
     }
   }
 
-  private CancellationTokenSource cancellationSource;
+  private CancellationTokenSource? cancellationSource;
 
   public void Cancel() {
-    if (cancellationSource == null) {
-      throw new InvalidOperationException("There is no ongoing run to cancel.");
-    }
-
-    cancellationSource.Cancel();
+    cancellationSource?.Cancel();
```

Comment:
The `x?.y` operator stores a reference to `x` before doing the null check, so even in a multi-thread scenario it doesn't throw a null reference exception.

Reply:
I wasn't thinking of that. I was thinking of calling cancel twice.

Reply:
What would be the issue there? I think the cancellation occurs once and that's it.

Reply:
More precisely, I found something here:
https://stackoverflow.com/a/52824946/1287856
So it's safe in your case, because you never call the `Dispose()` method of CancellationTokenSource.